pkg/controller/populators/import-populator.go

```diff
@@ -106,44 +106,66 @@ func (r *ImportPopulatorReconciler) Reconcile(_ context.Context, req reconcile.R
 // Implementations of populatorController methods
 
 // Import-specific implementation of getPopulationSource
-func (r *ImportPopulatorReconciler) getPopulationSource(namespace, name string) (client.Object, error) {
+func (r *ImportPopulatorReconciler) getPopulationSource(pvc *corev1.PersistentVolumeClaim) (client.Object, error) {
+	volumeImportSourceKey := types.NamespacedName{Namespace: getPopulationSourceNamespace(pvc), Name: pvc.Spec.DataSourceRef.Name}
 	volumeImportSource := &cdiv1.VolumeImportSource{}
-	volumeImportSourceKey := types.NamespacedName{Namespace: namespace, Name: name}
 	if err := r.client.Get(context.TODO(), volumeImportSourceKey, volumeImportSource); err != nil {
 		// reconcile will be triggered once created
 		if k8serrors.IsNotFound(err) {
 			return nil, nil
 		}
 		return nil, err
 	}
+	targetClaim := volumeImportSource.Spec.TargetClaim
+	if targetClaim != "" && targetClaim != pvc.Name {
+		r.log.V(1).Info("volumeImportSource is meant for a different PVC, ignoring")
+		return nil, nil
+	}
 	return volumeImportSource, nil
 }
 
 // Import-specific implementation of reconcileTargetPVC
 func (r *ImportPopulatorReconciler) reconcileTargetPVC(pvc, pvcPrime *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
 	pvcCopy := pvc.DeepCopy()
 	phase := pvcPrime.Annotations[cc.AnnPodPhase]
+	multiStageInProgress := false
+	source, err := r.getPopulationSource(pvc)
+	if source == nil {
+		return reconcile.Result{}, err
+	}
 
 	switch phase {
 	case string(corev1.PodRunning):
-		err := r.updatePVCWithPVCPrimeAnnotations(pvcCopy, pvcPrime, r.updateImportAnnotations)
+		if err = cc.MaybeSetPvcMultiStageAnnotation(pvcPrime, r.getCheckpointArgs(source)); err != nil {
+			return reconcile.Result{}, err
+		}
+		if err = r.updatePVCWithPVCPrimeAnnotations(pvcCopy, pvcPrime, r.updateImportAnnotations); err != nil {
+			return reconcile.Result{}, err
+		}
 		// We requeue to keep reporting progress
-		return reconcile.Result{RequeueAfter: 2 * time.Second}, err
+		return reconcile.Result{RequeueAfter: 2 * time.Second}, nil
 	case string(corev1.PodFailed):
 		// We'll get called later once it succeeds
 		r.recorder.Eventf(pvc, corev1.EventTypeWarning, importFailed, messageImportFailed, pvc.Name)
 	case string(corev1.PodSucceeded):
+		if _, multiStageInProgress = pvc.Annotations[cc.AnnCurrentCheckpoint]; multiStageInProgress {
```

Comment:
maybe put this in a function since you use it twice and it will be cleaner?

Reply:
Sure.

Reply:
Had to fix a bug involving this annotation so ended up following your suggestion.